pageserver/src/page_service.rs

```diff
@@ -3312,69 +3356,453 @@ where
     }
 }
 
-/// Implements the page service over gRPC.
+/// Serves the page service over gRPC. Dispatches to PageServerHandler for request processing.
+///
+/// TODO: rename to PageServiceHandler when libpq impl is removed.
+pub struct GrpcPageServiceHandler {
+    tenant_manager: Arc<TenantManager>,
+    ctx: RequestContext,
+}
+
+impl GrpcPageServiceHandler {
+    /// Errors if the request is executed on a non-zero shard. Only shard 0 has a complete view of
+    /// relations and their sizes, as well as SLRU segments and similar data.
+    #[allow(clippy::result_large_err)]
+    fn ensure_shard_zero(timeline: &Handle<TenantManagerTypes>) -> Result<(), tonic::Status> {
+        match timeline.get_shard_index().shard_number.0 {
+            0 => Ok(()),
+            shard => Err(tonic::Status::invalid_argument(format!(
+                "request must execute on shard zero (is shard {shard})",
+            ))),
+        }
+    }
+
+    /// Generates a PagestreamRequest header from a ReadLsn and request ID.
+    fn make_hdr(read_lsn: page_api::ReadLsn, req_id: u64) -> PagestreamRequest {
+        PagestreamRequest {
+            reqid: req_id,
+            request_lsn: read_lsn.request_lsn,
+            not_modified_since: read_lsn
+                .not_modified_since_lsn
+                .unwrap_or(read_lsn.request_lsn),
+        }
+    }
+
+    /// Acquires a timeline handle for the given request.
+    ///
+    /// TODO: consider moving this to a middleware layer; all requests need it. Needs to manage
+    /// the TimelineHandles lifecycle.
+    ///
+    /// TODO: untangle acquisition from TenantManagerWrapper::resolve() and Cache::get(), to avoid
+    /// the unnecessary overhead.
+    async fn get_request_timeline(
+        &self,
+        req: &tonic::Request<impl Any>,
+    ) -> Result<Handle<TenantManagerTypes>, GetActiveTimelineError> {
+        let ttid = *extract::<TenantTimelineId>(req);
+        let shard_index = *extract::<ShardIndex>(req);
+        let shard_selector = ShardSelector::Known(shard_index);
```

Comment:
How is this going to work with concurrent shard splits?

Let's say you have a request for shard `0000` and we are splitting into `0002` and `0102`.
On the pageserver side, we create the new shards and then shut-down `0000`.
The compute notification will be sent _after_ this process is complete.

This worked well with the previous approach because shard 0 remains on the same node
and moving it would trigger a compute notification, but now that the compute specifies
the shard, it will be unavailable until the compute notification is delivered.

Reply:
> The working plan is to temporarily install some kind of shard routing from the parent shard to child shards on the same Pageserver during the split, and remove it once the compute has been notified.

I think that should be fine. If you get a miss here, you can just look up any number zero shard on the node.

Reply:
We'll have to look up non-zero shards, because requests would come in for the parent. But the parent isn't necessarily shard zero, because we do repeated splits now. We can find child shards by looking for shards with larger shard counts.